File: 0279-perfect-squares/0279-perfect-squares.py
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [float('inf')] * (n + 1)
        dp[0] = 0

        count = 1
        sq = count * count

        while sq <= n:
            for i in range(sq, n + 1):
                dp[i] = min(dp[i - sq] + 1, dp[i])
            count += 1
            sq = count * count

        return dp[n]

File: 0279-perfect-squares/0279-perfect-squares.ts
function numSquares(n: number): number {
    const dp: number[] = new Array(n + 1).fill(Infinity);
    dp[0] = 0;

    let count = 1;
    let sq = count * count;

    while (sq <= n) {
        for (let i = sq; i <= n; i++) {
            dp[i] = Math.min(dp[i - sq] + 1, dp[i]);
        }
        count++;
        sq = count * count;
    }

    return dp[n];
}
